src/algebra/periodic.lean

```diff
@@ -0,0 +1,381 @@
+/-
+Copyright (c) 2021 Benjamin Davidson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Benjamin Davidson
+-/
+import data.int.parity
+import algebra.module.opposites
+
+/-!
+# Periodicity
+
+In this file we define and then prove facts about periodic and antiperiodic functions.
+
+## Main definitions
+
+* `function.periodic`: A function `f` is *periodic* if `∀ x, f (x + c) = f x`.
+  `f` is referred to as periodic with period `c` or `c`-periodic.
+
+* `function.antiperiodic`: A function `f` is *antiperiodic* if `∀ x, f (x + c) = -f x`.
+  `f` is referred to as antiperiodic with antiperiod `c` or `c`-antiperiodic.
+
+Note that any `c`-antiperiodic function will necessarily also be `2*c`-periodic.
+
+## Tags
+
+period, periodic, periodicity, antiperiodic
+-/
+
+variables {α β γ : Type*} {f g : α → β} {c c₁ c₂ x : α}
+
+namespace function
+
+/-! ### Periodicity -/
+
+/-- A function `f` is said to be `periodic` with period `c` if for all `x`, `f (x + c) = f x`. -/
+@[simp] def periodic [has_add α] (f : α → β) (c : α) : Prop :=
+∀ x : α, f (x + c) = f x
+
+lemma periodic.comp [has_add α]
+  (h : periodic f c) (g : β → γ) :
+  periodic (g ∘ f) c :=
+by simp * at *
+
```

Comment:
Do you think this lemma is worth having too?

```suggestion

lemma periodic.comp_add_hom [has_add α] [has_add γ]
  (h : periodic f c) (g : add_hom γ α) (g_inv : α → γ) (hg : function.right_inverse g_inv g) :
  periodic (f ∘ g) (g_inv c) :=
λ x, by simp [periodic, hg c, h (g x)]

```

Reply:
I don't know much about homomorphisms so I can't really say; if you think it's potentially useful I will certainly add it.

Reply:
(The same goes for any other versions you can think of)